File: rustbridge-java/rustbridge-core/src/main/java/com/rustbridge/LifecycleState.java
package com.rustbridge;

import org.jetbrains.annotations.NotNull;

/**
 * Plugin lifecycle states.
 * <p>
 * The lifecycle follows this state machine:
 * <pre>
 * Installed → Starting → Active → Stopping → Stopped
 *                ↑                    │
 *                └────────────────────┘ (restart)
 *            Any state → Failed (on error)
 * </pre>
 */
public enum LifecycleState {
    /**
     * Plugin is installed but not yet initialized.
     */
    INSTALLED(0),

    /**
     * Plugin is starting up.
     */
    STARTING(1),

    /**
     * Plugin is active and ready to handle requests.
     */
    ACTIVE(2),

    /**
     * Plugin is shutting down.
     */
    STOPPING(3),

    /**
     * Plugin has been stopped.
     */
    STOPPED(4),

    /**
     * Plugin has failed.
     */
    FAILED(5);

    private final int code;

    LifecycleState(int code) {
        this.code = code;
    }

    /**
     * Get the state from a numeric code.
     *
     * @param code the state code
     * @return the corresponding state
     * @throws IllegalArgumentException if the code is invalid
     */
    public static @NotNull LifecycleState fromCode(int code) {
        for (LifecycleState state : values()) {
            if (state.code == code) {
                return state;
            }
        }
        throw new IllegalArgumentException("Invalid state code: " + code);
    }

    /**
     * Get the numeric code for this state.
     *
     * @return the state code
     */
    public int getCode() {
        return code;
    }

    /**
     * Check if the plugin can handle requests in this state.
     *
     * @return true if requests can be handled
     */
    public boolean canHandleRequests() {
        return this == ACTIVE;
    }

    /**
     * Check if this is a terminal state.
     *
     * @return true if the plugin cannot transition to other states
     */
    public boolean isTerminal() {
        return this == STOPPED || this == FAILED;
    }
}

File: rustbridge-csharp/RustBridge.Core/LifecycleState.cs
namespace RustBridge;

/// <summary>
/// Plugin lifecycle states.
/// <para>
/// The lifecycle follows this state machine:
/// <code>
/// Installed -> Starting -> Active -> Stopping -> Stopped
///                 ^                      |
///                 +----------------------+ (restart)
///             Any state -> Failed (on error)
/// </code>
/// </para>
/// </summary>
public enum LifecycleState
{
    /// <summary>
    /// Plugin is installed but not yet initialized.
    /// </summary>
    Installed = 0,

    /// <summary>
    /// Plugin is starting up.
    /// </summary>
    Starting = 1,

    /// <summary>
    /// Plugin is active and ready to handle requests.
    /// </summary>
    Active = 2,

    /// <summary>
    /// Plugin is shutting down.
    /// </summary>
    Stopping = 3,

    /// <summary>
    /// Plugin has been stopped.
    /// </summary>
    Stopped = 4,

    /// <summary>
    /// Plugin has failed.
    /// </summary>
    Failed = 5
}

/// <summary>
/// Extension methods for <see cref="LifecycleState"/>.
/// </summary>
public static class LifecycleStateExtensions
{
    /// <summary>
    /// Get the state from a numeric code.
    /// </summary>
    /// <param name="code">The state code.</param>
    /// <returns>The corresponding state.</returns>
    /// <exception cref="ArgumentException">If the code is invalid.</exception>
    public static LifecycleState FromCode(int code)
    {
        if (Enum.IsDefined(typeof(LifecycleState), code))
        {
            return (LifecycleState)code;
        }
        throw new ArgumentException($"Invalid state code: {code}", nameof(code));
    }

    /// <summary>
    /// Check if the plugin can handle requests in this state.
    /// </summary>
    /// <param name="state">The lifecycle state.</param>
    /// <returns>True if requests can be handled.</returns>
    public static bool CanHandleRequests(this LifecycleState state) => state == LifecycleState.Active;

    /// <summary>
    /// Check if this is a terminal state.
    /// </summary>
    /// <param name="state">The lifecycle state.</param>
    /// <returns>True if the plugin cannot transition to other states.</returns>
    public static bool IsTerminal(this LifecycleState state) =>
        state == LifecycleState.Stopped || state == LifecycleState.Failed;
}
